ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -84,35 +126,72 @@ public boolean uploadBlob(BlobId blobId, long blobSize, InputStream blobInputStr
         return false;
       }
 
-      //azureBlob.setMetadata(new HashMap<>());
+      azureBlob.setMetadata(getMetadataMap(cloudBlobMetadata));
       azureBlob.upload(blobInputStream, blobSize, null, options, opContext);
+
+      documentClient.createDocument(cosmosCollectionLink, cloudBlobMetadata, defaultRequestOptions, true);
+
       LOGGER.debug("Uploaded blob {} to Azure container {}.", blobId, azureContainer.getName());
       return true;
-    } catch (URISyntaxException | StorageException | IOException e) {
+    } catch (URISyntaxException | StorageException | DocumentClientException | IOException e) {
       throw new CloudStorageException("Failed to upload blob: " + blobId, e);
     }
   }
 
   @Override
-  public boolean deleteBlob(BlobId blobId) throws CloudStorageException {
-    Objects.requireNonNull(blobId);
+  public boolean deleteBlob(BlobId blobId, long deletionTime) throws CloudStorageException {
+    return updateBlobMetadata(blobId, CloudBlobMetadata.FIELD_DELETION_TIME, deletionTime);
+  }
+
+  @Override
+  public boolean updateBlobExpiration(BlobId blobId, long expirationTime) throws CloudStorageException {
+    return updateBlobMetadata(blobId, CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime);
+  }
+
+  /**
+   * Update the metadata for the specified blob.
+   * @param blobId The {@link BlobId} to update.
+   * @param fieldName The metadata field to modify.
+   * @param value The new value.
+   * @return {@code true} if the udpate succeeded, {@code false} if the metadata record was not found.
+   * @throws DocumentClientException
+   */
+  private boolean updateBlobMetadata(BlobId blobId, String fieldName, Object value) throws CloudStorageException {
+    Objects.requireNonNull(blobId, "BlobId cannot be null");
+    Objects.requireNonNull(fieldName, "Field name cannot be null");
+
+    // We update the blob deletion time in two places:
+    // 1) the CosmosDB metadata collection
+    // 2) the blob storage entry metadata (to enable rebuilding the database)
 
     try {
       CloudBlobContainer azureContainer = getContainer(blobId, false);
       CloudBlockBlob azureBlob = azureContainer.getBlockBlobReference(blobId.getID());
 
       if (!azureBlob.exists()) {
-        LOGGER.debug("Skipping deletion of blob {} as it does not exist in Azure container {}.", blobId,
-            azureContainer.getName());
+        LOGGER.debug("Blob {} not found in Azure container {}.", blobId, azureContainer.getName());
         return false;
       }
-
-      // TODO: set deletedTime metadata
-      azureBlob.delete();
-      LOGGER.debug("Deleted blob {} from Azure container {}.", blobId, azureContainer.getName());
+      azureBlob.downloadAttributes(); // Makes sure we have latest
+      azureBlob.getMetadata().put(fieldName, String.valueOf(value));
```

Comment:
Open comment: is it useful to add an extra filed called `originalTTL`?

Reply:
I know we have that in blob properties but don't see much use for it in cloud metadata.